### DISCUSSION TITLE
How to customize existing projects

### DISCUSSION THREAD

**Discussion by ITler** [projen/projen#2283] [Q&A] [answered]
Hello again,

I want to use the `TypescriptProject` from projen as a base to build my own customization upon it.
I figured out, that i.e. parts of the eslint config (`.eslintrc.json`) does not match my needs. I would need to **remove** some parts of it (i.e. the `extends` section contents) and add my own bits of pieces. However the current API only supports **add**ing config.

So I tried to go with deleting the whole file and tried to build the file from scratch. Unfortunately, projen forces me to also remove the `eslint` task, which never was my intention. However, doing so projen forces me to remove the `test` task because of some dependencies... the goes on and I end up with:

```typescript
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: 'main',
  name: 'template-ts-base',
  projenrcTs: true,
  github: false,
  sampleCode: false,
});

project.removeTask('release');
project.removeTask('build');
project.removeTask('test');
project.removeTask('eslint');

// because I am not happy with the provided config
project.tryRemoveFile(".eslintrc.json")

const eslint = new javascript.Eslint(project, {
  dirs: []
});

eslint.addExtends("my contents here")
project.synth();
```

How do I get back the original implementation for the tasks?
Or is there a better approach? Or let's say? What is your intention of how the projen is to be used?

**Comment by mrgrain** [comment]
> I figured out, that i.e. parts of the eslint config (.eslintrc.json) does not match my needs.

You are probably better off using the `project. tryFindObjectFile('.eslintrc.json')` API.

>  Unfortunately, projen forces me to also remove the eslint task, which never was my intention. 

What do you mean by "projen forcing you"? I see you've added `project.removeTask('eslint');`, but why did you?

> How do I get back the original implementation for the tasks?

I don't think that's possible, aside from copying the relevant bits of code.

--- 

> Or is there a better approach? Or let's say? What is your intention of how the projen is to be used?

Currently it's tricky to extend existing projects. Like you've discovered, we are missing manipulation APIs everywhere. PRs are welcome!

In general, projen (and jsii) tend to prefer the declarative way of defining everything as props/options. But that makes retro fitting difficult. For wider changes, I'd probably recommend to completely disable `eslint` completely (pass `eslint: false` to the project constructor) and then start from there. There's also `eslintOptions` in the constructor that allows you to set a lot of things.

I'm not exactly sure what you can't get with the existing component, but I'd have expected that you can configure everything with the [Eslint component](https://projen.io/api/API.html#projen-javascript-eslint). If anything isn't possible, let's fix that.

**Reply by ITler** [reply]
ah yes, the `eslint: false` I missed, that led to all the other modifications I did. Alright got it. Thanks for your support

**Comment by mrgrain** [comment, marked as answer]
There is a tool now for exactly this use case:
https://github.com/mrgrain/jsii-struct-builder

**Reply by ITler** [reply]
thanks for sharing
